accord-core/src/main/java/accord/impl/SimpleProgressLog.java

```diff
@@ -101,332 +90,343 @@ static Progress advance(Progress current)
-                        globalProgress = Expected;
-                    }
-                    refreshGlobal(node, command, node.id(), null);
+                    status = CoordinateStatus.Done;
+                    progress = NoneExpected;
                 case Done:
             }
         }
 
-        void updateLocal(Node node, TxnId txnId, Command command)
+        void update(Node node, CommandStore commandStore, TxnId txnId, Command command)
         {
-            if (localProgress != NoProgress)
+            if (progress != NoProgress)
             {
-                localProgress = advance(localProgress);
+                progress = advance(progress);
                 return;
             }
 
-            localProgress = Investigating;
-            switch (local)
+            progress = Investigating;
+            switch (status)
             {
                 default: throw new IllegalStateException();
                 case NotWitnessed:
                 case Committed:
+                case Done:
                     throw new IllegalStateException(); // NoProgressExpected
 
                 case Uncommitted:
                 case ReadyToExecute:
                 {
-                    if (local.isAtLeast(Committed) && global.isAtLeast(PendingDurable))
+                    if (status.isAtLeast(CoordinateStatus.Committed) && command.durability().isDurable())
                     {
                         // must also be committed, as at the time of writing we do not guarantee dissemination of Commit
                         // records to the home shard, so we only know the executeAt shards will have witnessed this
                         // if the home shard is at an earlier phase, it must run recovery
-                        Key homeKey = command.homeKey();
-                        long homeEpoch = command.executeAt().epoch;
-
-                        node.withEpoch(homeEpoch, () -> {
-                            Shard homeShard = node.topology().forEpoch(homeKey, homeEpoch);
-                            debugInvestigating = checkOnCommitted(node, txnId, homeKey, homeShard, command.executeAt().epoch)
-                                                 .addCallback((success, fail) -> {
-                                                     // should have found enough information to apply the result, but in case we did not reset progress
-                                                     if (localProgress == Investigating)
-                                                         localProgress = Expected;
-                                                 });
-                        });
+                        long epoch = command.executeAt().epoch;
+                        node.withEpoch(epoch, () -> FetchData.fetch(Apply, node, txnId, command.route(), epoch, (success, fail) -> {
+                            // should have found enough information to apply the result, but in case we did not reset progress
```

Comment:
If it loops forever, it means there's a bug, and we probably want that (as we can open up the debugger and see where it's looping, and why). Since this implementation is only for toy / test / canonical purposes that's fine. I'll add a comment to this effect.